### PR TITLE
feat/Autocomplete Generics

### LINKED ISSUE #1188: Use generics for types where possible
### Describe what feature you'd like. Pseudo-code, mockups, or screenshots of similar solutions are encouraged!

Initially thought of this for the `Autocomplete` component, but after [discussion on Discord](https://discord.com/channels/1003691521280856084/1003699523522142399/1087811314526720091), it seems this may be a more sweeping change.

[Svelte has support for generic types](https://github.com/sveltejs/rfcs/pull/38). We can constrain them similar to `<T extends S>` in a generic:
```ts
type GenericType = $$Generic<{value: string}>;
export let selected: GenericType | undefined;
```
Now if we had a list of items of type `B = { value: string;  foo: Bar }`, and the user sets `bind:selected` on the parent, the type of `selected` would be `B`, instead of just `{value: string}`

### What type of pull request would this be?

Enhancement

### Any links to similar examples or other references we should review?

_No response_

### PR DESCRIPTION
**Pull request by HugeLetters**
## Linked Issue

Closes #1188 

## Description

Fixes generics on values and other props in Autocomplete component

## Changsets

feat: autocomplete generics

Instructions: Changesets automate our changelog. If you modify files in `/packages/skeleton`, run `pnpm changeset` in the root of the monorepo, follow the prompts, then commit the markdown file. Changes that add features should be `minor` while chores and bugfixes should be `patch`. Please prefix the changeset message with `feat:`, `bugfix:` or `chore:`.

## Checklist

Please read and apply all [contribution requirements](https://www.skeleton.dev/docs/contributing).

- [x] This PR targets the `dev` branch (NEVER `master`)
- [x] Documentation reflects all relevant changes
- [x] Branch is prefixed with: `docs/`, `feat/`, `chore/`, `bugfix/`
- [x] Ensure Svelte and Typescript linting is current - run `pnpm check`
- [x] Ensure Prettier linting is current - run `pnpm format`
- [x] All test cases are passing - run `pnpm test`
- [x] Includes a changeset (if relevant; see above)

### CODE REVIEW

File: .vscode/settings.json
Comment (HugeLetters) [inline .vscode/settings.json:1]: I had to add this - otherwise my default Prettier settings would change formatting

**Comment by endigo9740** [comment]
Thanks for submitting the draft @HugeLetters. @AdrianGonz97 and @ryceg are our resident Typescript experts, so I'm going to ping them here to provide feedback asap.

To make sure we're on the same page, I [mentioned this in the main thread](https://github.com/skeletonlabs/skeleton/issues/1188#issuecomment-1712139388) this might be something we want to extend to Modals and other Skeleton features too. I'm fine if we want to take a piecemeal approach to that, along as these aren't lost in the shuffle.

File: .vscode/settings.json
Comment (endigo9740) [reply .vscode/settings.json:1]: @niktek I'll let you review this specifically

Comment (niktek) [reply .vscode/settings.json:1]: How did you format ? The formatting should be done via package script i.e. `pnpm format` which uses the project settings and not VSCode.  Put another way, 2.8k commits in we haven't needed those settings before.

Comment (HugeLetters) [reply .vscode/settings.json:1]: > How did you format ? The formatting should be done via package script i.e. `pnpm format` which uses the project settings and not VSCode. Put another way, 2.8k commits in we haven't needed those settings before.

I have a format on save turned on - so either I need to disable it in workspace settings which would also go to that commit or I need to change my default Prettier settings. 
So for example I can't check-out your script files w/o formatting them.

**Comment by HugeLetters** [comment]
> Thanks for submitting the draft @HugeLetters.
> 
> @AdrianGonz97 and @ryceg are our resident Typescript experts, so I'm going to ping them here to provide feedback asap.
> 
> To make sure we're on the same page, I [mentioned in the main thread](https://github.com/skeletonlabs/skeleton/issues/1188#issuecomment-1712139388) this approach might be something we want to extend to Modals and other Skeleton features too. I'm fine if we want to take a piecemeal approach to that, as long as these aren't lost in the shuffle.

Sure, once this gets approved and I get a better understanding of kind of code style is appreciated here I'll try to work on other components too. 
Btw, what should the procedure be on minor improvements? Like [here](https://github.com/HugeLetters/skeleton/blob/620aaafa697043fd099c7abc3585e1b0b12b546c/packages/skeleton/src/lib/components/Autocomplete/Autocomplete.svelte#L98) there's an unnecessary array copying or [here](https://github.com/HugeLetters/skeleton/blob/620aaafa697043fd099c7abc3585e1b0b12b546c/packages/skeleton/src/lib/components/Autocomplete/Autocomplete.svelte#L155) when there's no limit set.

**Comment by endigo9740** [comment]
> Btw, what should the procedure be on minor improvements? ...

We keep scope small and limited in PRs. This helps keep changes localized and not grow too large. For additional issues it's worth creating new issues and dedicated PRs for each change.

I will also advise you to browser our issue backlog. There's a number of issues folks have with the autocomplete component specifically. It's a feature implemented by me, but not one I've had a lot of prior experience with, so I think there's definitely some room for improvement. At this point we're reaching a point a more consorted effort to improve the feature top to bottom might be in order.

If you're not already on Discord, I might recommend joining and jumping into the `#contributor` channel. This is a great place to discuss and plan bigger ideas. That or creating a Discussions section. Once a set plan is established we can generate one or more issues against that.

**Review by AdrianGonz97 — CHANGES_REQUESTED** [review]
I love the proposed changes, great job! I pushed a couple of changes when it came to annotating types and some variable names.

One thing I wanted to mention is that since we're already modifying the type of `value` into a generic, we may as well do the same for `meta`. 

Additionally, `keywords` looks to accept a `string` type, despite being labeled as `any`. Note the comment for it: `/** Provide a comma separated list of keywords. */`. What are your thoughts on this @endigo9740? I looked back in the original PR that introduced the component and didn't see any conversations about keeping `keywords` as `any`, so I'm assuming this is an `any` type that slipped between the cracks.

**Comment by endigo9740** [comment]
@AdrianGonz97 Per `keywords`, yes those should be a `string` ideally. Offhand I can't recall if we're parsing those, but they will be read as a string in the end.

**Comment by AdrianGonz97** [comment]
@endigo9740 Thanks! I thought as much. Looking at the code, we're not parsing it at all and the only time it's used is when everything is being strinigified for search - see [here](https://github.com/skeletonlabs/skeleton/blob/f7a4d2331815b9bb59b0c0adf1d6edd7ef09eeee/packages/skeleton/src/lib/components/Autocomplete/Autocomplete.svelte#L121).

In that case, I think it's safe to change that `any` to a `string` then.

File: .vscode/settings.json
Comment (HugeLetters) [reply .vscode/settings.json:1]: > Can't say that I'm a fan of this particular change. I think there's a better way to go about this rather than having 2 sources of truth for prettier. I have an idea for this, but for now I'll suggest reverting this change from the PR and we can address this in a separate PR.

Sure, I can just not include these settings changes in my PR - wasn't sure about the best way to approach this but a formatting config enforced directly by your repo would be nice I think.

**Comment by HugeLetters** [comment]
> I love the proposed changes, great job! I pushed a couple of changes when it came to annotating types and some variable names.
> 
> One thing I wanted to mention is that since we're already modifying the type of `value` into a generic, we may as well do the same for `meta`.
> 
> Additionally, `keywords` looks to accept a `string` type, despite being labeled as `any`. Note the comment for it: `/** Provide a comma separated list of keywords. */`. What are your thoughts on this @endigo9740? I looked back in the original PR that introduced the component and didn't see any conversations about keeping `keywords` as `any`, so I'm assuming this is an `any` type that slipped between the cracks.

Working on making Meta generic.

Also a bit of a problem with Value generic are scenarios like this
image
Clearly here I would input to be the source of truth for the generic but TS thinks options takes precedence - I honestly have no idea if it's possible to somehow enforce this behavior. But I will say I don't think it's anything more than annoyance since you can avoid this with explicit typing.
Specifically this error goes away if you just move input to a variable instead of inline(as you should).

And also I have to update your code in the docs like this after adding a generic for Meta
```ts
type FlavorOption = AutocompleteOption<string, { healthy: boolean }>;
const flavorOptions: FlavorOption[] = [
	{ label: 'Vanilla', value: 'vanilla', keywords: 'plain, basic', meta: { healthy: false } },
	{ label: 'Chocolate', value: 'chocolate', keywords: 'dark, white', meta: { healthy: false } },
	{ label: 'Strawberry', value: 'strawberry', keywords: 'fruit', meta: { healthy: true } },
	{ label: 'Neapolitan', value: 'neapolitan', keywords: 'mix, strawberry, chocolate, vanilla', meta: { healthy: false } },
	{ label: 'Pineapple', value: 'pineapple', keywords: 'fruit', meta: { healthy: true } },
	{ label: 'Peach', value: 'peach', keywords: 'fruit', meta: { healthy: true } }
];
```
Because otherwise TS infers generic Meta type as `{healthy: false}` and then gives error on every `{health: true}` instead of inferring it as a union.

File: packages/skeleton/src/lib/components/Autocomplete/Autocomplete.svelte
Comment (HugeLetters) [inline packages/skeleton/src/lib/components/Autocomplete/Autocomplete.svelte:25]: Would love a suggestion if there's a better name for this, haha. 
Since this is reused so much throughout the component I've extracted it to a separate type.

File: packages/skeleton/src/lib/components/Autocomplete/types.ts
Comment (HugeLetters) [inline packages/skeleton/src/lib/components/Autocomplete/types.ts:11]: A DX improvement could be made that if meta value is provided then it can't be undefined like so
```ts
export type AutocompleteOption<Value = unknown, Meta = unknown> = {
	/** Provide a unique display label per option. Supports HTML. */
	label: string;
	/** Provide a unique option value. */
	value: Value;
	/** Provide a comma separated list of keywords. */
	keywords?: string;
} & ([undefined] extends [Meta]
	? {
			/** Pass arbitrary data per option. */
			meta?: Meta;
	  }
	: {
			/** Pass arbitrary data per option. */
			meta: Meta;
	  });
```

Not sure it that's necessary though and even if it is - if there's a nicer way to do that. Don't think there's a way to do it with interfaces.

File: sites/skeleton.dev/src/routes/(inner)/components/autocomplete/+page.svelte
Comment (HugeLetters) [inline sites/skeleton.dev/src/routes/(inner)/components/autocomplete/+page.svelte]: Had to change this because otherwise TS would infer `Meta` generic type as `{healthy: false}` instead of `boolean` and then would error that other meta tags do not satisfy the type.

**Comment by endigo9740** [comment]
Hey @AdrianGonz97 not sure if you saw but @HugeLetters had a few follow up questions above. If we can get these addressed and give me a thumbs up when this is approved and ready to merge.

File: packages/skeleton/src/lib/components/Autocomplete/Autocomplete.svelte
Comment (AdrianGonz97) [reply packages/skeleton/src/lib/components/Autocomplete/Autocomplete.svelte:25]: Given that the value it represents is named `options`, I think this is perfectly fine 😄

File: packages/skeleton/src/lib/components/Autocomplete/types.ts
Comment (AdrianGonz97) [reply packages/skeleton/src/lib/components/Autocomplete/types.ts:11]: I don't think that'll be necessary since it'll likely be a breaking change for those that vary the type of their `meta` field. 

For instance, consider the following:
```ts
// note that we're not specifying a Meta type, so it defaults to `unknown`
type FlavorOption = AutocompleteOption<string>;
const flavorOptions: FlavorOption[] = [
	{ label: 'Vanilla', value: 'vanilla', keywords: 'plain, basic', meta: { healthy: false } },
	{ label: 'Chocolate', value: 'chocolate', keywords: 'dark, white', meta: { healthy: false } },
	{ label: 'Peach', value: 'peach', keywords: 'fruit' }
];
```
The last value in the array (Peach) will have a type error, despite `meta` being optional. It can be fixed by specifying the type for Meta in `AutocompleteOption`, but the default behavior will break.

Ideally, we'd want the default generic type of `Meta` to be `undefined`, but that's also going to result in breaking the types for people that have already implemented the `AutocompleteOption` type.

So for now, let's just keep it as is and I'll note that it should be changed to `undefined` in the next major version of Skeleton.

**Review by AdrianGonz97 — APPROVED** [review]
Excellent changes and thanks for the update! 

I've pushed a small change that adjusts the type for `options` so that it's appearance in the docs site makes a little more sense. And with that, everything else looks great and I think this PR is ready to be merged.

**Comment by HugeLetters** [comment]
> Excellent changes and thanks for the update!
> 
> I've pushed a small change that adjusts the type for `options` so that it's appearance in the docs site makes a little more sense. And with that, everything else looks great and I think this PR is ready to be merged.

Great! Somebody needs to authorize Vercel's deployment to follow through.

**Pull request merged**
